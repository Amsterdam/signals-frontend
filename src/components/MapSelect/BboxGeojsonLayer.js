const BboxGeojsonLayer = L.GeoJSON.extend({
  options: {
    zoomMin: 0,
    zoomMax: 99
  }, // Default options

  //
  // Leaflet layer methods
  //
  initialize(extraOptions, options) {
    L.GeoJSON.prototype.initialize.call(this, undefined, options);
    L.Util.setOptions(this, extraOptions);

    // istanbul ignore next
    if (!this.options.fetchRequest) {
      throw new Error('missing fetchRequest option');
    }
  },

  onAdd(map) {
    this._map = map;

    map.on('moveend', this.onMoveEnd, this);
    map.on('zoomend', this.onZoomEnd, this);
    map.on('refresh', this.onRefresh, this);

    this.fetchNewData();
  },

  onRemove(map) {
    map.off('moveend', this.onMoveEnd, this);
    map.off('zoomend', this.onZoomEnd, this);
    map.off('refresh', this.onRefresh, this);

    // Remove any geometry on this layer
    L.LayerGroup.prototype.onRemove.call(this, map);

    this._map = null;
  },

  //
  // Custom methods
  //
  zoomInRange() {
    const zoom = this._map.getZoom();
    return zoom >= this.options.zoomMin && zoom <= this.options.zoomMax;
  },

  fetchNewData() {
    if (!this.zoomInRange()) {
      // Outside zoom range, not fetching new data
      return;
    }

    this.isLoading = true;
    this.fire('loading');

    const bounds = this._map.getBounds();
    this.options.fetchRequest(bounds.toBBoxString())
      .then((geoData) => {
        this.clearLayers(); // Remove previous geometry
        this.addData(geoData);  // Adds geojson object to this layer

        this.isLoading = false;
        this.fire('load');
      })
      .catch(() => {
        this.isLoading = false;
        this.fire('error');
      });
  },

  onMoveEnd() {
    // Fired after dragging AND zooming!
    this.fetchNewData();
  },

  onZoomEnd() {
    if (!this.zoomInRange()) {
      this.clearLayers(); // Remove previous geometry, new data is fetched by onMoveEnd
    }
  },

  onRefresh() {
    this.fetchNewData();
  },
});

export default function (extraOptions, options) {
  return new BboxGeojsonLayer(extraOptions, options);
}
