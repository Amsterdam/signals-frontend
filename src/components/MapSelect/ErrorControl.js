const ErrorControl = L.Control.extend({
  options: {
    message: 'Oops, something went wrong'
  },

  initialize: function(options) {
    L.setOptions(this, options);
  },

  onAdd: function(map) {
    const div = L.DomUtil.create('div', 'error-control');
    div.innerText = this.options.message;
    L.DomUtil.addClass(div, 'hide');
    return div;
  },

  show: function() {
    L.DomUtil.removeClass(this._container, 'hide');
  },

  hide: function() {
    L.DomUtil.addClass(this._container, 'hide');
  }
});

const constructor = function(opts) {
  return new ErrorControl(opts);
};

export default constructor;

