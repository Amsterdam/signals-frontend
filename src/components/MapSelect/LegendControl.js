const LegendControl = L.Control.extend({
  options: {
    elements: []
  },

  initialize(options) {
    L.setOptions(this, options);
  },

  onAdd() {
    const div = L.DomUtil.create('div', 'legend-control');

    const header = L.DomUtil.create('div', 'legend-header', div);
    header.innerText = 'Legenda';

    const content = L.DomUtil.create('div', 'legend-content', div);
    this.options.elements.forEach((element) => {
      const item = L.DomUtil.create('div', 'legend-content-item', content);
      const icon = L.DomUtil.create('img', 'legend-content-icon', item);
      icon.src = element.iconUrl;
      const text = L.DomUtil.create('span', 'legend-content-text', item);
      text.innerText = element.label;
    });

    return div;
  }
});

const constructor = (opts) => new LegendControl(opts);
export default constructor;
