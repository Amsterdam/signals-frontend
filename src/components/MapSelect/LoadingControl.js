/**
 * Loading message component using Leaflet Control API.
 * NOTE, known limitation: not properly handling parallel loading events. First success will hide element.
 */
const LoadingControl = L.Control.extend({
  options: {
  },

  initialize: function(options) {
    L.setOptions(this, options);
  },

  onAdd: function(map) {
    this.addLayerListeners(map);

    const div = this.options.element;
    this.checkVisiblity(div);
    return div;
  },

  onRemove: function(map) {
    this.removeLayerListeners(map);
  },

  //
  // Custom methods
  //
  inFlight: [],

  onDataLoading: function (event) {
    const id = event.target._leaflet_id;
    this.inFlight.push(id);
    this.checkVisiblity(this._container);
  },
  onDataLoad: function (event) {
    const id = event.target._leaflet_id;
    this.inFlight = this.inFlight.filter(item => item !== id);
    this.checkVisiblity(this._container);
  },

  checkVisiblity: function(element) {
    const isVisible = this.inFlight.length > 0;
    if (isVisible) {
      L.DomUtil.removeClass(element, 'hide');
    } else {
      L.DomUtil.addClass(element, 'hide');
    }
  },

  addLayerLoadListener: function(layer) {
    if (!layer || !layer.on) return;
    layer.on({
      loading: this.onDataLoading,
      load: this.onDataLoad
    }, this);
  },

  removeLayerLoadListener: function(layer) {
    if (!layer || !layer.on) return;
    layer.off({
      loading: this.onDataLoading,
      load: this.onDataLoad
    }, this);
  },

  onLayerAdd: function(event) {
    this.addLayerLoadListener(event.layer);
  },

  onLayerRemove: function(event) {
    this.removeLayerLoadListener(event.layer);
  },

  addLayerListeners: function(map) {
    map.eachLayer(layer => this.addLayerLoadListener(layer));

    map.on({
      layeradd: this.onLayerAdd,
      layerremove: this.onLayerRemove
    }, this);
  },

  removeLayerListeners: function(map) {
    map.eachLayer(layer => this.removeLayerLoadListener(layer));

    map.off({
      layeradd: this.onLayerAdd,
      layerremove: this.onLayerRemove
    }, this);
  },

});

const constructor = function(opts) {
  return new LoadingControl(opts);
};

export default constructor;

