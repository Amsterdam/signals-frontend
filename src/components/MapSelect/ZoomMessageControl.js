const ZoomMessageControl = L.Control.extend({
  options: {
    zoomMin: 0
  },

  initialize: function(options) {
    L.setOptions(this, options);
  },

  onAdd: function(map) {
    const div = L.DomUtil.create('div');
    div.innerText = "Zoom in om de objecten te zien";
    L.DomUtil.setClass(div, 'zoom-message');

    this.setVisibility(div, map.getZoom());

    map.on('zoomend', this.onZoomEnd, this);

    return div;
  },

  onRemove: function(map) {
    map.off('zoomend', this.onZoomEnd, this);
  },

  //
  // Custom methods
  //
  onZoomEnd: function () {
    this.setVisibility(this._container, this._map.getZoom());
  },

  setVisibility: function (element, zoomLevel) {
    if (zoomLevel >= this.options.zoomMin) {
      L.DomUtil.addClass(element, 'hide');
    } else {
      L.DomUtil.removeClass(element, 'hide');
    }
  }
});

const constructor = function(opts) {
  return new ZoomMessageControl(opts);
};

export default constructor;

