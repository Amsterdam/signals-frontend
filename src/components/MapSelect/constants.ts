export const ZOOM_MIN = 13
export const ZOOM_MAX = 15
