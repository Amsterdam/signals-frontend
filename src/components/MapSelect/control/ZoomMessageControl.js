const ZoomMessageControl = L.Control.extend({
  options: {
    zoomMin: 0,
  },

  initialize(options) {
    L.setOptions(this, options);
  },

  onAdd(map) {
    const div = L.DomUtil.create('div', 'zoom-control');
    div.innerText = 'Zoom in om de objecten te zien';

    this.setVisibility(div, map.getZoom());

    map.on('zoomend', this.onZoomEnd, this);

    return div;
  },

  onRemove(map) {
    map.off('zoomend', this.onZoomEnd, this);
  },

  //
  // Custom methods
  //
  onZoomEnd() {
    this.setVisibility(this._container, this._map.getZoom());
  },

  setVisibility(element, zoomLevel) {
    if (zoomLevel >= this.options.zoomMin) {
      L.DomUtil.addClass(element, 'hide');
    } else {
      L.DomUtil.removeClass(element, 'hide');
    }
  },
});

export default ZoomMessageControl;
