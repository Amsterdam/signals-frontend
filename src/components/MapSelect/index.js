import React, { useLayoutEffect, useEffect, useState, useCallback, useMemo, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import BboxGeojsonLayer from '@datapunt/leaflet-geojson-bbox-layer';
import 'leaflet/dist/leaflet.css';
import classNames from 'classnames';

import Map from 'components/Map';
import MAP_OPTIONS from 'shared/services/configuration/map-options';
import request from 'utils/request';
import MaxSelection from 'utils/maxSelection';

import ZoomMessageControl from './control/ZoomMessageControl';
import LegendControl from './control/LegendControl';
import LoadingControl from './control/LoadingControl';
import ErrorControl from './control/ErrorControl';

import './style.scss';

const SELECTION_MAX_COUNT = 30;

const Wrapper = styled.div`
  position: relative;
`;

const StyledMap = styled(Map)`
  height: 450px;
  width: 100%;
  font-family: "Avenir Next LT W01-Regular", arial, sans-serif;
`;

const MapSelect = ({
  geojsonUrl,
  getIcon,
  hasGPSControl,
  iconField,
  idField,
  latlng,
  legend,
  onSelectionChange,
  selectionOnly,
  value,
}) => {
  const zoomMin = 13;
  const featuresLayer = useRef();
  const [mapInstance, setMapInstance] = useState();
  const selection = useRef(new MaxSelection(SELECTION_MAX_COUNT, value));
  const mapOptions = {
    ...MAP_OPTIONS,
    center: [latlng.latitude, latlng.longitude],
    zoomControl: false,
    minZoom: 10,
    maxZoom: 15,
    zoom: 14,
  };

  const errorControl = useMemo(
    () =>
      new ErrorControl({
        position: 'topleft',
        message: 'Oops, de objecten kunnen niet worden getoond. Probeer het later nog eens.',
      }),
    []
  );

  const fetchRequest = useCallback(
    bbox_str =>
      request(`${geojsonUrl}&bbox=${bbox_str}`).catch(() => {
        errorControl.show();
      }),
    [errorControl, geojsonUrl]
  );

  const bboxGeoJsonLayer = useMemo(
    () =>
      BboxGeojsonLayer(
        {
          fetchRequest,
        },
        {
          zoomMin,

          zoomMax: 15,

          /**
           * Function that will be used to decide whether to include a feature or not. The default is to include all
           * features.
           *
           * Note that this behaviour is difficult to test, hence the istanbul ignore
           */
          filter: /* istanbul ignore next */ feature =>
            selectionOnly ? selection.current.has(feature.properties[idField]) : true,

          /**
           * Function defining how GeoJSON points spawn Leaflet layers. It is internally called when data is added,
           * passing the GeoJSON point feature and its LatLng.
           * Return value overridden to have it return a marker with a specific icon
           *
           * Note that this behaviour is difficult to test, hence the istanbul ignore
           */
          pointToLayer: /* istanbul ignore next */ (feature, latlong) =>
            L.marker(latlong, {
              icon: getIcon(feature.properties[iconField], selection.current.has(feature.properties[idField])),
              alt: feature.properties.objectnummer,
            }),

          /**
           * Function called once for each created Feature, after it has been created and styled.
           * Attaches click handler to markers that are rendered on the map
           *
           * Note that this behaviour is difficult to test, hence the istanbul ignore
           */
          onEachFeature: /* istanbul ignore next */ (feature, layer) => {
            if (onSelectionChange) {
              // Check that the component is in write mode
              layer.on({
                click: e => {
                  const _layer = e.target;
                  const id = _layer.feature.properties[idField];
                  selection.current.toggle(id);

                  onSelectionChange(selection.current);
                },
              });
            }
          },
        }
      ),
    [fetchRequest, getIcon, iconField, idField, onSelectionChange, selection, selectionOnly]
  );

  /**
   * Set the features layer on mount
   */
  useLayoutEffect(() => {
    featuresLayer.current = bboxGeoJsonLayer;
    // only execute on mount; disabling linter
    // eslint-disable-next-line
  }, []);

  /**
   * Initialise the whole map when its instance can be retrieved from the DOM
   */
  useEffect(() => {
    if (!mapInstance) return undefined;

    featuresLayer.current.addTo(mapInstance);

    const zoomMessageControl = new ZoomMessageControl({
      position: 'topleft',
      zoomMin,
    });

    zoomMessageControl.addTo(mapInstance);

    if (legend) {
      // only show if legend items are provided
      const legendControl = new LegendControl({
        position: 'topright',
        zoomMin,
        elements: legend,
      });

      legendControl.addTo(mapInstance);
    }

    const div = L.DomUtil.create('div', 'loading-control');
    div.innerText = 'Bezig met laden...';

    const loadingControl = new LoadingControl({
      position: 'topleft',
      element: div,
    });

    loadingControl.addTo(mapInstance);

    errorControl.addTo(mapInstance);

    return () => {
      mapInstance.remove();
    };
    // only execute when the mapInstance is available; disabling linter
    // eslint-disable-next-line
  }, [mapInstance]);

  /**
   * Registering to value changes
   * Value changes happen when a marker on the map is clicked (see onEachFeature). Each marker is looped over and its
   * correct icon is set based on the selected values.
   *
   * Note that this behaviour is next to impossible to test. Hence the istanbul ignore.
   */
  useEffect(
    /* istanbul ignore next */ () => {
      if (!featuresLayer.current) return;

      selection.current.set.clear();

      for (const id of value) {
        selection.current.add(id);
      }

      // Let icons reflect new selection
      featuresLayer.current.getLayers().forEach(layer => {
        const properties = layer.feature.properties;
        const id = properties[idField];
        const iconType = properties[iconField];
        const icon = getIcon(iconType, selection.current.has(id));

        layer.setIcon(icon);
      });
    },
    // only execute when value changes; disabling linter
    // eslint-disable-next-line
    [value]
  );

  return (
    <Wrapper>
      <StyledMap
        className={classNames('map-component', { write: onSelectionChange })}
        data-testid="mapSelect"
        hasGPSControl={hasGPSControl}
        hasZoomControls
        mapOptions={mapOptions}
        setInstance={setMapInstance}
      />
    </Wrapper>
  );
};

MapSelect.defaultProps = {
  hasGPSControl: false,
  value: [],
  selectionOnly: false,
};

MapSelect.propTypes = {
  latlng: PropTypes.exact({
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
  }).isRequired,
  geojsonUrl: PropTypes.string.isRequired,
  onSelectionChange: PropTypes.func,
  getIcon: PropTypes.func.isRequired,
  hasGPSControl: PropTypes.bool,
  legend: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      iconUrl: PropTypes.string.isRequired,
    })
  ),
  iconField: PropTypes.string.isRequired,
  idField: PropTypes.string.isRequired,
  value: PropTypes.arrayOf(PropTypes.string),
  selectionOnly: PropTypes.bool,
};

export default MapSelect;
