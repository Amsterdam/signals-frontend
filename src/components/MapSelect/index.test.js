import React from 'react';
import { render } from '@testing-library/react';

import { withAppContext } from 'test/utils';
import ZoomMessageControl from './control/ZoomMessageControl';
import LegendControl from './control/LegendControl';
import LoadingControl from './control/LoadingControl';
import ErrorControl from './control/ErrorControl';

import MapSelect from './index';

jest.mock('./control/ZoomMessageControl');
jest.mock('./control/LegendControl');
jest.mock('./control/LoadingControl');
jest.mock('./control/ErrorControl');

const fetchResponse = {
  type: 'FeatureCollection',
  name: 'klokken',
  crs: { type: 'name', properties: { name: 'urn:ogc:def:crs:OGC:1.3:CRS84' } },
  features: [
    {
      type: 'Feature',
      properties: { ogc_fid: '48634', type_id: '1', type_name: 'Klok', objectnummer: '065121' },
      geometry: { type: 'Point', coordinates: [4.883614, 52.37855] },
    },
    {
      type: 'Feature',
      properties: { ogc_fid: '93331', type_id: '1', type_name: 'Klok', objectnummer: '017431' },
      geometry: { type: 'Point', coordinates: [4.877793, 52.379065] },
    },
  ],
};

fetch.mockResponse(JSON.stringify(fetchResponse));

describe('<MapSelect />', () => {
  const legend = [{ key: 'klok', label: 'Klok', iconUrl: 'foo/bar/icon.svg' }];
  const onSelectionChange = jest.fn();
  const url = 'foo/geo.json?';
  const getIcon = (type, isSelected) => {
    if (isSelected) {
      return L.divIcon({ className: 'my-div-icon-select' });
    }

    return L.divIcon({ className: 'my-div-icon' });
  };
  const latlng = {
    latitude: 4,
    longitude: 52,
  };

  beforeEach(() => {
    jest.resetAllMocks();
    fetch.resetMocks();
  });

  it('should render correctly', async () => {
    const { findByTestId, getByTestId } = render(
      withAppContext(
        <MapSelect
          latlng={latlng}
          onSelectionChange={onSelectionChange}
          getIcon={getIcon}
          geojsonUrl={url}
          iconField="type_name"
          idField="objectnummer"
          hasGPSControl
        />
      )
    );

    await findByTestId('map-base');

    expect(getByTestId('gpsButton')).toBeInTheDocument();
    expect(LegendControl).not.toHaveBeenCalled();
    expect(ZoomMessageControl.mock.instances[0].addTo).toHaveBeenCalled();
    expect(ErrorControl.mock.instances[0].addTo).toHaveBeenCalled();
    expect(LoadingControl.mock.instances[0].addTo).toHaveBeenCalled();
  });

  it('should render legend', async () => {
    const { findByTestId } = render(
      withAppContext(
        <MapSelect
          latlng={latlng}
          onSelectionChange={onSelectionChange}
          getIcon={getIcon}
          geojsonUrl={url}
          legend={legend}
          iconField="type_name"
          idField="objectnummer"
        />
      )
    );

    await findByTestId('map-base');

    expect(LegendControl).toHaveBeenCalled();
  });

  it('should do bbox fetch', async () => {
    expect(fetch).not.toHaveBeenCalled();

    const { findByTestId } = render(
      withAppContext(
        <MapSelect
          latlng={latlng}
          onSelectionChange={onSelectionChange}
          getIcon={getIcon}
          legend={legend}
          geojsonUrl={url}
          iconField="type_name"
          idField="objectnummer"
        />
      )
    );

    await findByTestId('map-base');

    const bboxRegex = /bbox=(\d{1,2}\.\d{1,16},?){4}$/;
    expect(fetch).toHaveBeenCalledWith(expect.stringMatching(bboxRegex), undefined);
  });
});
