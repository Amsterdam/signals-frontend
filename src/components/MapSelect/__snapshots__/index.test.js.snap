// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MapSelect /> should render correctly 1`] = `
<div
  className="map-component write"
>
  <div
    className="map"
  >
    <div
      className="map-container"
      id="mapSelect"
    />
  </div>
</div>
`;
